libcudacxx/include/cuda/std/inplace_vector

```diff
@@ -0,0 +1,1808 @@
+    auto __first = __range.begin();
+    auto __last  = __range.end();
+    _CCCL_IF_CONSTEXPR (_Capacity == 0)
+    {
+      if (__first != __last)
+      {
+        _CUDA_VSTD::__throw_bad_alloc();
+      }
+      return nullptr;
+    }
+
+    const iterator __old_end = this->end();
+    for (; __first != __last; ++__first)
+    {
+      emplace_back(*__first);
+    }
+
+    const iterator __pos = (iterator) __cpos;
+    _CUDA_VSTD::rotate(__pos, __old_end, this->end());
+    return __pos;
+  }
+
+  _LIBCUDACXX_TEMPLATE(class _Range)
+  _LIBCUDACXX_REQUIRES(
+    _CUDA_VRANGES::__container_compatible_range<_Range, _Tp> _LIBCUDACXX_AND _CUDA_VRANGES::forward_range<_Range>)
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr iterator insert_range(const_iterator __cpos, _Range&& __range)
+  {
+    auto __first = __range.begin();
+    _CCCL_IF_CONSTEXPR (_Capacity == 0)
+    {
+      if (__first != __range.end())
+      {
+        _CUDA_VSTD::__throw_bad_alloc();
+      }
+      return nullptr;
+    }
+
+    return insert(__cpos, __first, __unwrap_end(__range));
+  }
+
+  _LIBCUDACXX_TEMPLATE(class _Range)
+  _LIBCUDACXX_REQUIRES(
+    _CUDA_VRANGES::__container_compatible_range<_Range, _Tp> _LIBCUDACXX_AND(!_CUDA_VRANGES::forward_range<_Range>))
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr void append_range(_Range&& __range)
+  {
+    auto __first = __range.begin();
+    auto __last  = __range.end();
+    for (; __first != __last; ++__first)
+    {
+      emplace_back(*__first);
+    }
+  }
+
+  _LIBCUDACXX_TEMPLATE(class _Range)
+  _LIBCUDACXX_REQUIRES(
+    _CUDA_VRANGES::__container_compatible_range<_Range, _Tp> _LIBCUDACXX_AND _CUDA_VRANGES::forward_range<_Range>)
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr void append_range(_Range&& __range)
+  {
+    auto __first = __range.begin();
+    _CCCL_IF_CONSTEXPR (_Capacity == 0)
+    {
+      if (__first != __range.end())
+      {
+        _CUDA_VSTD::__throw_bad_alloc();
+      }
+      return;
+    }
+
+    insert(this->end(), __first, __unwrap_end(__range));
+  }
+#  endif // _CCCL_STD_VER >= 2017 && !defined(_CCCL_COMPILER_MSVC_2017)
+
+  template <class... _Args>
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr iterator emplace(const_iterator __cpos, _Args&&... __args)
+  {
+    const iterator __pos = (iterator) __cpos;
+    const iterator __end = this->end();
+    if (this->size() == _Capacity)
+    {
+      _CUDA_VSTD::__throw_bad_alloc();
+    }
+    else if (__pos < this->begin() || __end < __pos)
+    {
+      _CUDA_VSTD::__throw_out_of_range("inplace_vector::emplace(const_iterator, Args...)");
+    }
+
+    if (__pos == __end)
+    {
+      this->unchecked_emplace_back(_CUDA_VSTD::forward<_Args>(__args)...);
+    }
+    else
+    {
+      _Tp __temp{_CUDA_VSTD::forward<_Args>(__args)...};
+      this->unchecked_emplace_back(_CUDA_VSTD::move(*(__end - 1)));
+      _CUDA_VSTD::move_backward(__pos, __end - 1, __end);
+      *__pos = _CUDA_VSTD::move(__temp);
```

Comment:
Question: The early construction of `__temp` is for strong exception safety guarantee in case the constructor of `_Tp` throws, right?

Reply:
Yes